lms/lmsweb/git_service.py

```diff
@@ -0,0 +1,219 @@
+
+    @property
+    def repository_folder(self) -> pathlib.Path:
+        return pathlib.Path(self._base_repository_folder) / self.project_name
+
+    def handle_operation(self) -> flask.Response:
+        git_operation = self._extract_git_operation()
+        repository_folder = self.repository_folder / 'config'
+
+        new_repository = not repository_folder.exists()
+        if new_repository:
+            self._initialize_bare_repository()
+
+        if not git_operation.supported:
+            raise OSError
+
+        data_out = self._execute_git_operation(git_operation)
+
+        if git_operation.format_response:
+            data_out = git_operation.format_response(data_out)
+
+        if git_operation.contain_new_commits:
+            files = self._load_files_from_repository()
+            solution_hash = hashing.by_content(str(files))
+            upload.upload_solution(
+                course_id=self._course_id,
+                exercise_number=self._exercise_number,
+                files=files,
+                solution_hash=solution_hash,
+                user_id=self._user.id,
+            )
+
+        return self.build_response(data_out, git_operation)
+
+    def _execute_command(
+            self,
+            args: typing.List[str],
+            cwd: typing.Union[str, pathlib.Path],
+            proc_input: typing.Optional[bytes] = None,
+    ):
+        proc = subprocess.Popen(  # noqa: S603
+            args=args,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            cwd=cwd,
+        )
+        data_out, _ = proc.communicate(proc_input, self._GIT_PROCESS_TIMEOUT)
+        operation_failed = proc.wait() != self._GIT_VALID_EXIT_CODE
+        if operation_failed:
+            log.error(
+                'Failed to execute command %s. stdout=%s\nstderr=%s',
+                args, proc.stdout.read(), proc.stderr.read(),
+            )
+            raise OSError
+        return data_out
+
+    def _execute_git_operation(self, git_operation: _GitOperation) -> bytes:
+        return self._execute_command(
+            args=git_operation.service_command,
+            cwd=self._base_repository_folder,
+            proc_input=self._request.data,
+        )
+
+    def _initialize_bare_repository(self) -> None:
+        os.makedirs(self.repository_folder, exist_ok=True)
+        self._execute_command(
+            args=['git', 'init', '--bare'],
+            cwd=self.repository_folder,
+        )
+
+    @staticmethod
+    def build_response(
+            data_out: bytes,
+            git_operation: _GitOperation,
+    ) -> flask.Response:
+        res = flask.make_response(data_out)
+        res.headers['Pragma'] = 'no-cache'
+        res.headers['Cache-Control'] = 'no-cache, max-age=0, must-revalidate'
+        res.headers['Content-Type'] = git_operation.response_content_type
+        return res
+
+    def _extract_git_operation(self) -> _GitOperation:
```

Comment:
WDYT about splitting it to shorter functions and return the variables as a dictionary?